src/QsCompiler/SyntaxProcessor/ExpressionVerification.fs

```diff
@@ -396,19 +396,23 @@ let private characteristicsSet info =
         | Controlled -> Controllable)
     |> Set.ofSeq
 
-let private lambdaCharacteristics (body: TypedExpression) =
-    let mutable characteristics = Set.ofList [ Adjointable; Controllable ]
-
-    let onCall =
-        function
+let private lambdaCharacteristics (inference: InferenceContext) (body: TypedExpression) =
+    let mutable characteristics =
+        if inference.Resolve(body.ResolvedType).Resolution = UnitType then
+            Set.ofList [ Adjointable; Controllable ]
+        else
+            Set.empty
+
+    let onCall callableType =
```

Comment:
This function is a bit difficult to understand without any comments. The operation match case took me a bit, but I think I understand that that is used to collect up all the characteristic sets from all operation expressions in the lambda body and get the minimal characteristic set from among them. Is that right? Either way, could you please add a comment explaining this for readability?

Reply:
I added a few comments. Do they help?

Reply:
Yes, thanks!